### ISSUE TITLE
Let `upstreamExchange` depend on `bodyType`

### ISSUE THREAD

**Issue by DDvO** [siemens/cmp-ra-component#2]
As discussed, to cope with EJBCA limitations on the authorization of KUR and RR messages,
we should allow that the upstream interface depends not only on the certProfile:
```
   CmpRaUpstream(final PersistencyContextManager persistencyContextManager,
            final Configuration config,
            final CmpBiFuncEx<PKIMessage, String, PKIMessage> upstreamExchange)
```
but also on the `bodyType`.

Unfortunately we had not foreseen this already at API design time.

Well, as a workaround the `upstreamExchange` function could peek into the ASN.1 encoding of the `PKIMessage` to find out the body type itself.
Yet this should better be done by the CmpRaComponentl because the outer application should not need to handle CMP message details.

**Comment by Vitowitsch** [comment]
This is also what we had in mind: A dedicated callback for us to redirect only to the KUR endpoint. Nice.

**Comment by Vitowitsch** [comment]
If you commit to this issue, we wait for it, instead of implementing the workaround.

**Comment by Akretsch** [comment]
The external accessible interface is:
```
public static final CmpRaInterface instantiateCmpRaComponent(
            final Configuration configuration,
            final BiFunction<byte[], String, byte[]> upstreamExchange)
            throws Exception;
```
_PersistencyContextManager_ , _CmpBiFuncEx_ and  _PKIMessage_ should not be used by the wrapping applications.

So the external interface may change to:

```
public static final CmpRaInterface instantiateCmpRaComponent(
            final Configuration configuration,
            final TriFunction<byte[], String, Integer, byte[]> upstreamExchange)
            throws Exception;
```
_TriFunction_ could be defined in _CmpRaComponent_ or come from _org.apache.commons.lang3.function_.

What's  the intended routing and handling of of the bodyType 24 (certConf)? Its used in imprinting, bootstrapping and key update.

**Comment by DDvO** [comment]
Good point that we need to be careful with the routing of certConf, because it may be different depending on whether it is in a ir/cr or kur. Actually the same issue is with pollReq.
I suggest that we define that for the selection of the upstreamExchange function the bodyType provided to the TriFunction is for subsequent messages in the same transaction always the bodyType of the first message.

**Comment by Akretsch** [comment]
implemented in f110c28196b645021a3f92dfea52c72515eb4fd4
